doc/man/man3/lis_output.3
.TH lis_output 3 "6 Sep 2012" "Man Page" "Lis Library Functions"

.SH NAME

lis_output \- write the matrix and vector data 

.SH SYNOPSIS

#include "lis.h"

\fBLIS_INT lis_output\fR(\fBLIS_MATRIX A\fR, \fBLIS_VECTOR b\fR, \fBLIS_VECTOR x\fR, \fBLIS_INT format\fI, \fBchar *filename\fR);

.SH DESCRIPTION

Write the matrix and vector data to the external file.

.SH INPUT

.IP "\fBA\fR"
The matrix

.IP "\fBb\fR"
The right hand side vector (If no vector is written to the external file, then \fBNULL\fR must be input.)

.IP "\fBx\fR"
The solution (If no vector is written to the external file, then \fBNULL\fR must be input.)

.IP "\fBformat\fR"
.RS
.IP "LIS_FMT_MM"
The Matrix Market format
.RE

.IP "\fBfilename\fR"
The destination file

.SH OUTPUT

.IP "\fBNone\fR"

.SH EXIT STATUS

The following exit values are returned:
.IP "0"
The process is normally terminated
.IP "unspecified"
An error occurred

.SH SEE ALSO

.BR lis (3)
.PP
http://www.ssisc.org/lis/



ChangeLog
Version 1.5.44  December 29, 2014

  - Fixed exception handling in matrix addition.
  - Updated documents.

Version 1.5.43  December 28, 2014

  - Fixed Fortran functions for matrix operations.

Version 1.5.42  December 27, 2014

  - Added support for matrix addition.

Version 1.5.41  December 23, 2014

  - Fixed bug in win/Makefile.in.

Version 1.5.40  December 20, 2014

  - Fixed variable declaration bug in lis_precision_vec.c.

Version 1.5.39  December 16, 2014

  - Updated documents.

Version 1.5.38  December 15, 2014

  - Updated documents.

Version 1.5.37  December 14, 2014

  - Cleaned up programs (thanks to Florian Lehner).

Version 1.5.36  December 11, 2014

  - Updated lis_esolver_rqi.c.

Version 1.5.35  December 10, 2014

  - Updated documents (thanks to Sushilkumar Sreekumar).

Version 1.5.34  December 8, 2014

  - Updated documents.
  - Changed default preconditioners for eigensolvers.

Version 1.5.33  December 7, 2014

  - Updated documents.
  - Added experimental support for Jacobi-Davidson.
  - Fixed exception handling bug in lis_matrix_set_size() (thanks to Yi Yao).

Version 1.5.32  December 5, 2014

  - Updated documents.

Version 1.5.31  December 4, 2014

  - Updated etest7.c.
  - Updated lis_esolver_ai.c.

Version 1.5.30  December 3, 2014

  - Updated documents.
  - Fixed print bug on iteration counts of inner eigensolvers.

Version 1.5.29  December 2, 2014

  - Added experimental support for Arnoldi iteration.
  - Disabled obsolete option -m for eigensolvers.
  - Updated documents.

Version 1.5.28  December 1, 2014

  - Changed default linear solver for eigensolvers.
  - Changed order of eigensolvers:

     flags      | eigensolvers              flags     | eigensolvers
    ------------+--------------------     ------------+--------------------
     -e {si|5}  | Subspace            ==>  -e {cg|5}  | CG
     -e {li|6}  | Lanczos                  -e {cr|6}  | CR
     -e {cg|7}  | CG                       -e {si|7}  | Subspace
     -e {cr|8}  | CR                       -e {li|8}  | Lanczos

  - Updated documents.

Version 1.5.27  November 30, 2014

  - Updated documents.

Version 1.5.26  November 29, 2014

  - Changed names of functions lis_array_matvec2() and
    lis_array_matmat2() to lis_array_matvec_ns() and lis_array_matmat_ns(),
    respectively.

Version 1.5.25  November 28, 2014

  - Updated lis_array.c.

Version 1.5.24  November 27, 2014

  - Updated lis_array.c (thanks to Florian Lehner).

Version 1.5.23  November 26, 2014

  - Cleaned up lis_esolver_li.c.
  - Cleaned up lis_array.c.
  - Fixed convergence condition of lis_array_qr().

Version 1.5.22  November 25, 2014

  - Updated documents.

Version 1.5.21  November 24, 2014

  - Cleaned up lis_solver_*.c.
  - Added support for test6f.F90 for array operations.
  - Updated documents.

Version 1.5.20  November 23, 2014

  - Updated documents.

Version 1.5.19  November 22, 2014

  - Fixed bug in lisf_array.c (thanks to Florian Lehner).

Version 1.5.18  November 21, 2014

  - Updated documents.

Version 1.5.17  November 20, 2014

  - Updated lis_fortran.h for Win64 environment.
  - Fixed initial value related bug in GMRES to enable option -initx_zeros.

Version 1.5.16  November 19, 2014

  - Updated documents (thanks to Sushilkumar Sreekumar).

Version 1.5.15  November 18, 2014

  - Fixed order of array elements in lis_array_cgs(), lis_array_mgs(),
    and lis_array_qr().
  - Fixed memory initialization bug in lis_array_mgs().
  - Updated documents.

Version 1.5.14  November 16, 2014

  - Cleaned up programs.
  - Updated eigensolvers.
  - Added sample scripts for graphics.

Version 1.5.13  November 14, 2014

  - Updated documents.

Version 1.5.12  November 13, 2014

  - Updated documents.

Version 1.5.11  November 12, 2014

  - Updated test3b.c.

Version 1.5.10  November 11, 2014

  - Added sample scripts for graphics.

Version 1.5.9  November 9, 2014

  - Cleaned up programs.

Version 1.5.8  November 8, 2014

  - Updated documents.

Version 1.5.7  November 7, 2014

  - Fixed print format.
  - Fixed elapsed time related bug in preconditioned eigensolvers.
  - Fixed CR related bug in lis_esolver_cg.c.

Version 1.5.6  November 6, 2014

  - Updated documents.

Version 1.5.5  November 5, 2014

  - Cleaned up programs.
  - Updated documents.

Version 1.5.4  November 4, 2014

  - Fixed quad precision related bug in lis_esolver_si.c.
  - Fixed bug in test/Makefile.am.

Version 1.5.3  November 3, 2014

  - Updated eigensolvers.
  - Fixed array related bug in lis_esolver_cg.c.

Version 1.5.2  November 1, 2014

  - Added support for etest7.c for array operations.
  - Changed specification of lis_array_qr() to return number of iterations
    and error of QR algorithm.
  - Updated documents.

Version 1.5.1  October 31, 2014

  - Added support for test6.c for array operations.
  - Updated print formats.
  - Updated documents.

Version 1.5.0  October 30, 2014

  - Fixed print bug in option -conv_cond (thanks to Suguru Miyauchi).
  - Added official support for array operations.
  - Updated documents.

Version 1.4.67  October 28, 2014

  - Fixed bug in lis_array_nrmi_f() (thanks to Florian Lehner).
  - Updated lis_array.c.

Version 1.4.66  October 27, 2014

  - Updated lis_array.c.
  - Changed names of functions lis_array_dot2() and
    lis_array_invGauss() to lis_array_dot() and lis_array_ge(),
    respectively.

Version 1.4.65  October 23, 2014

  - Updated documents (thanks to Suguru Miyauchi).

Version 1.4.64  October 21, 2014

  - Updated lis_m_data_creation_AMGCG.F90.
  - Updated sample scripts for graphics.

Version 1.4.63  October 19, 2014

  - Added sample scripts for graphics (thanks to Takashi Shiroto
    and Sushilkumar Sreekumar).

Version 1.4.62  October 17, 2014

  - Fixed exception handling bug in lis_m_data_creation_AMGCG.F90
    (thanks to Takashi Shiroto).

Version 1.4.61  September 19, 2014

  - Fixed OpenMP related bug in lis_matrix_ell.c (thanks to Gerald Wedel).

Version 1.4.60  September 13, 2014


src/system/lis_output.c
/* Copyright (C) 2005 The Scalable Software Infrastructure Project. All rights reserved.

   Redistribution and use in source and binary forms, with or without
   modification, are permitted provided that the following conditions are met:
   1. Redistributions of source code must retain the above copyright
      notice, this list of conditions and the following disclaimer.
   2. Redistributions in binary form must reproduce the above copyright
      notice, this list of conditions and the following disclaimer in the
      documentation and/or other materials provided with the distribution.
   3. Neither the name of the project nor the names of its contributors 
      may be used to endorse or promote products derived from this software 
      without specific prior written permission.

   THIS SOFTWARE IS PROVIDED BY THE SCALABLE SOFTWARE INFRASTRUCTURE PROJECT
   ``AS IS'' AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED
   TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR
   PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL THE SCALABLE SOFTWARE INFRASTRUCTURE
   PROJECT BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY,
   OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF
   SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS
   INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN
   CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE)
   ARISING IN ANY WAY OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE
   POSSIBILITY OF SUCH DAMAGE.
*/

#ifdef HAVE_CONFIG_H
	#include "lis_config.h"
#else
#ifdef HAVE_CONFIG_WIN_H
	#include "lis_config_win.h"
#endif
#endif

#include <stdio.h>
#include <stdlib.h>
#ifdef HAVE_MALLOC_H
        #include <malloc.h>
#endif
#include <string.h>
#include <stdarg.h>
#ifdef _OPENMP
	#include <omp.h>
#endif
#ifdef USE_MPI
	#include <mpi.h>
#endif
#include "lislib.h"

/************************************************
 * lis_output
 * lis_output_matrix
 * lis_output_vector
 * lis_output_vector_plain
 * lis_output_vector_mm
 * lis_output_vector_lis_ascii
 * lis_solver_output_rhistory
 * lis_esolver_output_rhistory
 ************************************************/

#undef __FUNC__
#define __FUNC__ "lis_output"
LIS_INT lis_output(LIS_MATRIX A, LIS_VECTOR b, LIS_VECTOR x, LIS_INT format, char *path)
{
	LIS_INT	err;
	LIS_MATRIX B;

	LIS_DEBUG_FUNC_IN;

	#ifdef USE_MPI
		MPI_Barrier(A->comm);
	#endif
	err = lis_matrix_check(A,LIS_MATRIX_CHECK_ALL);
	if( err ) return err;

	if( format==LIS_FMT_MM || format==LIS_FMT_MMB )
	{
		switch( A->matrix_type )
		{
		case LIS_MATRIX_CSR:
			err = lis_output_mm_csr(A,b,x,format,path);
			break;
		default:
			err = lis_matrix_duplicate(A,&B);
			if( err ) return err;
			lis_matrix_set_type(B,LIS_MATRIX_CSR);
			err = lis_matrix_convert(A,B);
			if( err ) return err;
			err = lis_output_mm_csr(B,b,x,format,path);
			lis_matrix_destroy(B);
			break;
		}
	}

	LIS_DEBUG_FUNC_OUT;
	return err;
}

#undef __FUNC__
#define __FUNC__ "lis_output"
LIS_INT lis_output_matrix(LIS_MATRIX A, LIS_INT format, char *path)
{
	LIS_INT err;
	LIS_MATRIX B;
	LIS_VECTOR b,x;

	LIS_DEBUG_FUNC_IN;

	#ifdef USE_MPI
		MPI_Barrier(A->comm);
	#endif
	err = lis_matrix_check(A,LIS_MATRIX_CHECK_ALL);
	if( err ) return err;

	err = lis_vector_create(LIS_COMM_WORLD,&b);
	err = lis_vector_create(LIS_COMM_WORLD,&x);

	if( format==LIS_FMT_MM || format==LIS_FMT_MMB )
	{
		switch( A->matrix_type )
		{
		case LIS_MATRIX_CSR:
			err = lis_output_mm_csr(A,b,x,format,path);
			break;
		default:
			err = lis_matrix_duplicate(A,&B);
			if( err ) return err;
			lis_matrix_set_type(B,LIS_MATRIX_CSR);
			err = lis_matrix_convert(A,B);
			if( err ) return err;
			err = lis_output_mm_csr(B,b,x,format,path);
			lis_matrix_destroy(B);
			break;
		}
	}

	err = lis_vector_destroy(b);
	err = lis_vector_destroy(x);

	LIS_DEBUG_FUNC_OUT;
	return err;
}

#undef __FUNC__
#define __FUNC__ "lis_output_vector"
LIS_INT lis_output_vector(LIS_VECTOR v, LIS_INT format, char *filename)
{
	LIS_INT err;

	LIS_DEBUG_FUNC_IN;

	#ifdef USE_MPI
		MPI_Barrier(v->comm);
	#endif
	err = lis_vector_check(v,LIS_VECTOR_CHECK_NULL);
	if( err ) return err;

	switch( format )
	{
	case LIS_FMT_PLAIN:
		err = lis_output_vector_plain(v,filename);
		break;
	case LIS_FMT_MM:
		err = lis_output_vector_mm(v,filename);
		break;
	case LIS_FMT_LIS:
		err = lis_output_vector_lis_ascii(v,filename);
		break;
	default:
		LIS_SETERR(LIS_ERR_ILL_ARG,"ill format option\n");
		return LIS_ERR_ILL_ARG;
		break;
	}

	LIS_DEBUG_FUNC_OUT;
	return err;
}

#undef __FUNC__
#define __FUNC__ "lis_output_vector_plain"
LIS_INT lis_output_vector_plain(LIS_VECTOR v, char *path)
{
	LIS_INT	n,i;
	#ifdef USE_MPI
		LIS_INT pe,nprocs,my_rank;
		LIS_INT	err,ret;
	#endif
	FILE *file;

	LIS_DEBUG_FUNC_IN;

#ifdef USE_MPI
	nprocs    = v->nprocs;
	my_rank   = v->my_rank;
	n         = v->n;

	err = 0;
	for(pe=0;pe<nprocs;pe++)
	{
		if( my_rank==pe )
		{
			if( my_rank==0 )
			{
				file = fopen(path, "w");
			}
			else
			{
				file = fopen(path, "a");
			}
			if( file==NULL )
			{
				LIS_SETERR1(LIS_ERR_FILE_IO,"cannot open file %s\n", path);
				err = 1;
			}
			for(i=0;i<n;i++)
			{
			  if (v->intvalue) 
			    {
#ifdef _LONGLONG
			      fprintf(file, "%28lld\n", (long long int)v->value[i]);
#else
			      fprintf(file, "%28d\n", (int)v->value[i]);
#endif
			    }
			  else
			    {
#ifdef _LONG__DOUBLE
				fprintf(file, "%28.20Le\n", v->value[i]);
#else
				fprintf(file, "%28.20e\n", v->value[i]);
#endif
			    }
			}
			fclose(file);
		}
		MPI_Allreduce(&err,&ret,1,LIS_MPI_INT,MPI_SUM,v->comm);
		if( ret )
		{
			return LIS_FAILS;
		}
	}
	LIS_DEBUG_FUNC_OUT;
	return LIS_SUCCESS;
#else
	n         = v->n;

	file = fopen(path, "w");
	if( file==NULL )
	{
		LIS_SETERR1(LIS_ERR_FILE_IO,"cannot open file %s\n", path);
		return LIS_ERR_FILE_IO;
	}
	for(i=0;i<n;i++)
	{
	  if (v->intvalue)
	    {
#ifdef _LONGLONG
	      fprintf(file, "%28lld\n", (long long int)v->value[i]);
#else
	      fprintf(file, "%28d\n", (int)v->value[i]);
#endif
	    }
	  else
	    {
#ifdef _LONG__DOUBLE
		fprintf(file, "%28.20Le\n", v->value[i]);
#else
		fprintf(file, "%28.20e\n", v->value[i]);
#endif
	    }
	}
	fclose(file);
	LIS_DEBUG_FUNC_OUT;
	return LIS_SUCCESS;
#endif
}

#undef __FUNC__
#define __FUNC__ "lis_output_vector_mm"
LIS_INT lis_output_vector_mm(LIS_VECTOR v, char *path)
{
	LIS_INT	i,n,is;
	#ifdef USE_MPI
		LIS_INT	pe,nprocs,my_rank;
		LIS_INT	err,ret;
	#endif
	FILE *file;

	LIS_DEBUG_FUNC_IN;
#ifdef USE_MPI
	nprocs    = v->nprocs;
	my_rank   = v->my_rank;
	n         = v->n;
	is        = v->is; 

	err = 0;
	if( my_rank==0 )
	{
		file = fopen(path, "w");
		if( file==NULL )
		{
			LIS_SETERR1(LIS_ERR_FILE_IO,"cannot open file %s\n", path);
			err = 1;
		}

		if (v->intvalue) 
		  {
		    fprintf(file, "%%%%MatrixMarket vector coordinate integer general\n");
		  }
		else
		  {
		    fprintf(file, "%%%%MatrixMarket vector coordinate real general\n");
		  }
#ifdef _LONGLONG
		fprintf(file, "%lld\n", v->gn);
#else
		fprintf(file, "%d\n", v->gn);
#endif
		fclose(file);
	}
	MPI_Allreduce(&err,&ret,1,LIS_MPI_INT,MPI_SUM,v->comm);
	if( ret )
	{
		return LIS_FAILS;
	}

	err = 0;
	for(pe=0;pe<nprocs;pe++)
	{
		if( my_rank==pe )
		{
			file = fopen(path, "a");
			if( file==NULL )
			{
				LIS_SETERR1(LIS_ERR_FILE_IO,"cannot open file %s\n", path);
				err = 1;
			}
			else
			{
				for(i=0;i<n;i++)
				{
				  if (v->intvalue)
				    {
#ifdef _LONGLONG
				      fprintf(file, "%lld %28lld\n", i+is+1, (long long int)v->value[i]); 
#else
				      fprintf(file, "%d %28d\n", i+is+1, (int)v->value[i]); 
#endif
				    }
				  else
				    {
#ifdef _LONGLONG
#ifdef _LONG__DOUBLE
					fprintf(file, "%lld %28.20Le\n", i+is+1, v->value[i]); 
#else
					fprintf(file, "%lld %28.20e\n", i+is+1, v->value[i]); 
#endif
#else
#ifdef _LONG__DOUBLE
					fprintf(file, "%d %28.20Le\n", i+is+1, v->value[i]); 
#else
					fprintf(file, "%d %28.20e\n", i+is+1, v->value[i]); 
#endif
#endif
				    }
				}
				fclose(file);
			}
		}
		MPI_Allreduce(&err,&ret,1,LIS_MPI_INT,MPI_SUM,v->comm);
		if( ret )
		{
			return LIS_FAILS;
		}
	}
	LIS_DEBUG_FUNC_OUT;
	return LIS_SUCCESS;
#else
	n         = v->n;
	is        = v->is; 

	file = fopen(path, "w");
	if( file==NULL )
	{
		LIS_SETERR1(LIS_ERR_FILE_IO,"cannot open file %s\n", path);
		return LIS_ERR_FILE_IO;
	}

	if (v->intvalue) 
	  {
	    fprintf(file, "%%%%MatrixMarket vector coordinate integer general\n");
	  }
	else
	  {
	    fprintf(file, "%%%%MatrixMarket vector coordinate real general\n");
	  }
#ifdef _LONGLONG
	fprintf(file, "%lld\n", v->gn);
#else
	fprintf(file, "%d\n", v->gn);
#endif

	for(i=0;i<n;i++)
	{
	  if (v->intvalue)
	    {
#ifdef _LONGLONG
	      fprintf(file, "%lld %28lld\n", i+is+1, (long long int)v->value[i]);
#else
	      fprintf(file, "%d %28d\n", i+is+1, (int)v->value[i]);
#endif
	    }
          else
	    {
#ifdef _LONGLONG
#ifdef _LONG__DOUBLE
	  fprintf(file, "%lld %28.20Le\n", i+is+1, v->value[i]);
#else
	  fprintf(file, "%lld %28.20e\n", i+is+1, v->value[i]);
#endif
#else
#ifdef _LONG__DOUBLE
	  fprintf(file, "%d %28.20Le\n", i+is+1, v->value[i]);
#else
	  fprintf(file, "%d %28.20e\n", i+is+1, v->value[i]);
#endif
#endif
	     }
	  /* fprintf(file, "%d %28.20e\n", i+1, v->value[i]); */
	}
	fclose(file);

	LIS_DEBUG_FUNC_OUT;
	return LIS_SUCCESS;
#endif
}


#undef __FUNC__
#define __FUNC__ "lis_output_vector_lis_ascii"
LIS_INT lis_output_vector_lis_ascii(LIS_VECTOR v, char *path)
{
	LIS_INT	i,n;
	#ifdef USE_MPI
		LIS_INT	pe,nprocs,my_rank;
		LIS_INT	err,ret;
	#endif
	FILE *file;

	LIS_DEBUG_FUNC_IN;
#ifdef USE_MPI
	nprocs    = v->nprocs;
	my_rank   = v->my_rank;
	n         = v->n;

	err = 0;
	if( my_rank==0 )
	{
		file = fopen(path, "w");
		if( file==NULL )
		{
			LIS_SETERR1(LIS_ERR_FILE_IO,"cannot open file %s\n", path);
			err = 1;
		}

		fprintf(file, "#LIS A vec\n");
#ifdef _LONGLONG
		fprintf(file, "%lld\n", nprocs);
#else
		fprintf(file, "%d\n", nprocs);
#endif
		fclose(file);
	}
	MPI_Allreduce(&err,&ret,1,LIS_MPI_INT,MPI_SUM,v->comm);
	if( ret )
	{
		return LIS_FAILS;
	}

	err = 0;
	for(pe=0;pe<nprocs;pe++)
	{
		if( my_rank==pe )
		{
			file = fopen(path, "a");
			if( file==NULL )
			{
				LIS_SETERR1(LIS_ERR_FILE_IO,"cannot open file %s\n", path);
				err = 1;
			}
			else
			{
#ifdef _LONGLONG
				fprintf(file, "# %lld %lld\n", pe, v->n);
#else
				fprintf(file, "# %d %d\n", pe, v->n);
#endif
				for(i=0;i<n;i++)
				{
				  if (v->intvalue) 
				    {
#ifdef _LONGLONG
				      fprintf(file, "%28lld ",(long long int)v->value[i]);
#else
				      fprintf(file, "%28d ",(int)v->value[i]);
#endif
				    }
				  else
				    {
#ifdef _LONG__DOUBLE
					fprintf(file, "%28.20Le ",v->value[i]);
#else
					fprintf(file, "%28.20e ",v->value[i]);
#endif
					if( (i+1)%3==0 ) fprintf(file, "\n");
				    }
				}
				if( n%3!=0 ) fprintf(file, "\n");
				fclose(file);
			}
		}
		MPI_Allreduce(&err,&ret,1,LIS_MPI_INT,MPI_SUM,v->comm);
		if( ret )
		{
			return LIS_FAILS;
		}
	}
	LIS_DEBUG_FUNC_OUT;
	return LIS_SUCCESS;
#else
	n         = v->n;

	file = fopen(path, "w");
	if( file==NULL )
	{
		LIS_SETERR1(LIS_ERR_FILE_IO,"cannot open file %s\n", path);
		return LIS_ERR_FILE_IO;
	}

	fprintf(file, "#LIS A vec\n");
	fprintf(file, "1\n");

#ifdef _LONGLONG
	fprintf(file, "# 0 %lld\n", v->n);
#else
	fprintf(file, "# 0 %d\n", v->n);
#endif
	for(i=0;i<n;i++)
	{
	  if (v->intvalue)
	    {
#ifdef _LONGLONG
	      fprintf(file, "%28lld ",(long long int)v->value[i]);
#else
	      fprintf(file, "%28d ",(int)v->value[i]);
#endif
	    }
	  else
	    {
#ifdef _LONG__DOUBLE
		fprintf(file, "%28.20Le ",v->value[i]);
#else
		fprintf(file, "%28.20e ",v->value[i]);
#endif
	    }
		if( (i+1)%3==0 ) fprintf(file, "\n");
	}
	if( n%3!=0 ) fprintf(file, "\n");
	fclose(file);

	LIS_DEBUG_FUNC_OUT;
	return LIS_SUCCESS;
#endif
}

#undef __FUNC__
#define __FUNC__ "lis_solver_output_rhistory"
LIS_INT lis_solver_output_rhistory(LIS_SOLVER solver, char *filename)
{
	LIS_INT	i,maxiter;
	#ifdef USE_MPI
		LIS_INT	my_rank;
	#endif
	FILE *file;

	LIS_DEBUG_FUNC_IN;

	maxiter = solver->iter+1;
	if( solver->retcode!=LIS_SUCCESS )
	{
		maxiter--;
	}
#ifdef USE_MPI
	if( solver->rhistory==NULL )
	{
		LIS_SETERR(LIS_FAILS,"residual history is empty\n");
		return LIS_FAILS;
	}
	if( solver->A==NULL )
	{
		LIS_SETERR(LIS_FAILS,"matrix A is NULL\n");
		return LIS_FAILS;
	}

	MPI_Barrier(solver->A->comm);
	my_rank = solver->A->my_rank;
	if( my_rank==0 )
	{
		file = fopen(filename, "w");
		if( file==NULL )
		{
			LIS_SETERR1(LIS_ERR_FILE_IO,"cannot open file %s\n", filename);
		}
		else
		{
			for(i=0;i<maxiter;i++)
			{
#ifdef _LONG__DOUBLE
				fprintf(file, "%Le\n", solver->rhistory[i]);
#else
				fprintf(file, "%e\n", solver->rhistory[i]);
#endif
			}
			fclose(file);
		}
	}
	MPI_Barrier(solver->A->comm);

	LIS_DEBUG_FUNC_OUT;
	return LIS_SUCCESS;
#else
	if( solver->rhistory==NULL )
	{
		LIS_SETERR(LIS_FAILS,"residual history is empty\n");
		return LIS_FAILS;
	}
	file = fopen(filename, "w");
	if( file==NULL )
	{
		LIS_SETERR1(LIS_ERR_FILE_IO,"cannot open file %s\n", filename);
		return LIS_ERR_FILE_IO;
	}
	for(i=0;i<maxiter;i++)
	{
#ifdef _LONG__DOUBLE
		fprintf(file, "%Le\n", solver->rhistory[i]);
#else
		fprintf(file, "%e\n", solver->rhistory[i]);
#endif
	}
	fclose(file);
	LIS_DEBUG_FUNC_OUT;
	return LIS_SUCCESS;
#endif
}

#undef __FUNC__
#define __FUNC__ "lis_esolver_output_rhistory"
LIS_INT lis_esolver_output_rhistory(LIS_ESOLVER esolver, char *filename)
{
	LIS_INT	i,maxiter;
	#ifdef USE_MPI
		LIS_INT	my_rank;
	#endif
	FILE *file;

	LIS_DEBUG_FUNC_IN;

	maxiter = esolver->iter[0]+1;
	if( esolver->retcode!=LIS_SUCCESS )
	{
		maxiter--;
	}
#ifdef USE_MPI
	if( esolver->rhistory==NULL )
	{
		LIS_SETERR(LIS_FAILS,"eigensolver's residual history is empty\n");
		return LIS_FAILS;
	}
	if( esolver->A==NULL )
	{
		LIS_SETERR(LIS_FAILS,"matrix A is NULL\n");
		return LIS_FAILS;
	}

	MPI_Barrier(esolver->A->comm);
	my_rank = esolver->A->my_rank;
	if( my_rank==0 )
	{
		file = fopen(filename, "w");
		if( file==NULL )
		{
			LIS_SETERR1(LIS_ERR_FILE_IO,"cannot open file %s\n", filename);
		}
		else
		{
			for(i=0;i<maxiter;i++)
			{
#ifdef _LONG__DOUBLE
				fprintf(file, "%Le\n", esolver->rhistory[i]);
#else
				fprintf(file, "%e\n", esolver->rhistory[i]);
#endif
			}
			fclose(file);
		}
	}
	MPI_Barrier(esolver->A->comm);

	LIS_DEBUG_FUNC_OUT;
	return LIS_SUCCESS;
#else
	if( esolver->rhistory==NULL )
	{
		LIS_SETERR(LIS_FAILS,"eigensolver's residual history is empty\n");
		return LIS_FAILS;
	}
	file = fopen(filename, "w");
	if( file==NULL )
	{
		LIS_SETERR1(LIS_ERR_FILE_IO,"cannot open file %s\n", filename);
		return LIS_ERR_FILE_IO;
	}
	for(i=0;i<maxiter;i++)
	{
#ifdef _LONG__DOUBLE
		fprintf(file, "%Le\n", esolver->rhistory[i]);
#else
		fprintf(file, "%e\n", esolver->rhistory[i]);
#endif
	}
	fclose(file);
	LIS_DEBUG_FUNC_OUT;
	return LIS_SUCCESS;
#endif
}

